File: 09-08-heap-y-cola-prio/adts/priority_queue/MinPriorityQueue.java
package adts.priority_queue;

import adts.heap.MinHeap;

public class MinPriorityQueue<E, P extends Comparable<P>> implements PriorityQueue<E, P> {
   private class Pair implements Comparable<Pair> {
      E elem;
      P prio;

      Pair() {
      }

      Pair(E elem, P prio) {
         this.elem = elem;
         this.prio = prio;
      }

      @Override
      public int compareTo(Pair other) {
         return this.prio.compareTo(other.prio);
      }
   }

   private MinHeap<Pair> h;

   public MinPriorityQueue(int expectedSize) {
      this.h = new MinHeap<>(expectedSize);
   }

   @Override
   public void push(E elem, P prio) {
      Pair p = new Pair(elem, prio);
      this.h.push(p);
   }

   @Override
   public E top() {
      Pair t = this.h.top();
      return t.elem;
   }

   @Override
   public E pop() {
      Pair t = this.h.pop();
      return t.elem;
   }

   @Override
   public int size() {
      return this.h.size();
   }

   @Override
   public boolean isEmpty() {
      return this.h.isEmpty();
   }

}

File: 09-08-heap-y-cola-prio/adts/priority_queue/min_priority_queue.cpp
#pragma once

#include "../heap/min_heap.cpp"
#include "priority_queue.cpp"
#include <cassert>

template <class E, class P>
class minPriorityQueue : public priorityQueue<E, P> {

private:
  struct pair {
    E elem;
    P prio;

    pair() {}

    pair(E elem, P prio) {
      this->elem = elem;
      this->prio = prio;
    }

    // this < other
    bool operator<(const pair &other) { return this->prio < other.prio; }
    bool operator>(const pair &other) { return this->prio > other.prio; }
    bool operator<=(const pair &other) { return this->prio <= other.prio; }
    bool operator>=(const pair &other) { return this->prio >= other.prio; }
    bool operator==(const pair &other) { return this->prio == other.prio; }
  };

  minHeap<pair> *h;

public:
  minPriorityQueue(int expectedSize) {
    this->h = new minHeap<pair>(expectedSize);
  }

  virtual void push(E elem, P prio) override {
    pair p = pair(elem, prio);
    this->h->push(p);
  }
  virtual E top() override {
    pair t = this->h->top();
    return t.elem;
  }

  virtual E pop() override {
    pair t = this->h->pop();
    return t.elem;
  }

  virtual int size() override { return this->h->size(); }
  virtual bool isEmpty() override { return this->h->isEmpty(); }
};
